src/librustc_typeck/check/wfcheck.rs

```diff
@@ -468,26 +469,55 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
 
         debug!("check_method_receiver: sig={:?}", sig);
 
+        let self_ty = fcx.normalize_associated_types_in(span, &self_ty);
+        let self_ty = fcx.liberate_late_bound_regions(
+            method.def_id,
+            &ty::Binder(self_ty)
+        );
+
         let self_arg_ty = sig.inputs()[0];
-        let rcvr_ty = match ExplicitSelf::determine(self_ty, self_arg_ty) {
-            ExplicitSelf::ByValue => self_ty,
-            ExplicitSelf::ByReference(region, mutbl) => {
-                fcx.tcx.mk_ref(region, ty::TypeAndMut {
-                    ty: self_ty,
-                    mutbl,
-                })
+
+        let cause = fcx.cause(span, ObligationCauseCode::MethodReceiver);
+        let eq = |expected, actual| fcx.at(&cause, fcx.param_env).eq(expected, actual);
```

Comment:
I think you want to wrap this inside a `commit_if_ok`, because IIRC just an `eq` isn't transactional.

Reply:
I don't know what you mean by transactional, but I'll try it; maybe it will help with the error messages problem

Reply:
@arielb1 on second thought, I had another approach [here](https://github.com/rust-lang/rust/pull/45870/commits/bbe755c7a6b20c1340ef1d86fb30d4b91b81450e#diff-15d40489387405cc3e23e2f9f66a74c2R494). If I did that instead, would it address your concern? (i.e. is `demand_eqtype_with_origin` "transactional"?)

Reply:
@mikeyhew 

Sorry forgot to detail - the problem is that `eq` can have side-effects even if it fails (e.g. if you equate `(_, u32)` with `(i32, i32)`, it will make the `_` be `i32` before erroring out). 

>  (i.e. is `demand_eqtype_with_origin` "transactional"?)

Nope. You want `fcx.commit_if_ok(|_| fcx.at(&cause, fcx.param_env).eq(expected, actual))`, which creates a transaction and rolls it back if the equality fails.